src/geometry.ts

```diff
@@ -125,20 +120,36 @@ function verticalLineIntersection(
   }
 }
 
-export function pointOnLineNearLine(
+export function pointOnPolylineNearPoint(p: Point, polyline: Point[]): Point {
+  const pointsOnSegments = pairs(polyline).map((segment) =>
+    pointOnLineNearPoint(...segment, p),
+  )
+
+  const pointOnNearestSegment = minBy(pointsOnSegments, (pt) =>
+    distanceBetweenPoints(p, pt),
+  )
+  if (!pointOnNearestSegment) {
+    throw new Error(`Not enough points on polyline`)
+  }
+
+  return pointOnNearestSegment[0]
+}
+
+function pointOnLineNearPoint(
   p1: Point,
   p2: Point,
   { x: x0, y: y0 }: Point,
 ): Point {
-  const a = p2.y - p1.y
-  const b = p1.x - p2.x
-  const c = p1.y * p2.x - p1.x * p2.y
-
-  const denom = a * a + b * b
-  const paren = a * y0 - b * x0
-  const x = (-b * paren - a * c) / denom
-  const y = (a * paren - b * c) / denom
-  return { x, y }
+  const d = Math.sqrt(Math.pow(p1.x - p2.x, 2) + Math.pow(p1.y - p2.y, 2))
```

Comment:
:exploding_head:

Reply:
and they said a B.Sc. in Mathematics wouldn't help me in life

Reply:
I'm impressed. :clap: